juju/juju
# Mongo snap only

There's 3 main changes here, mostly big deletions:
- mongo is now only installed from the juju-db snap
- server side transactions by default
- removal of apiserver and state juju restore code

The first commit is the removal of the cd from apiserver and state which existed to implement juju restore.
A lot of this code assumed mongo 3 etc and we needed to remove it anyway, so easier to do it now since mongo 3 support is also removed. Another side effect is we no longer need to track mongo version in agent.conf. And the restore watcher worker is also obsolete.

A necessary mgo test change is that repliacset support needs to be enabled on all mgo test suites, or else tests will fail on mongo 4.x.

The install of mongo on the controllers is now snap only, and 4.x only, meaning we can also turn on server side txns.
Todo is to figure out how to enable numa support for the snap.
Also todo is to delete juju restore CLI etc

## QA steps

```sh
juju bootstrap --bootstrap-series=xenial
juju status, juju deploy etc
```

check logs for errors

```sh
juju dump-db
juju enable-ha
juju show-controller # to check replicaset status
```

log into mongo and check `rs.config()`

```sh
juju create-backup
```

repeat on a focal controller also
